File: frontend/src/lib/types/model-usage.ts
export type ModelUsageSummary = {
  modelCalls: number;
  confirmedCalls: number;
  failedCalls: number;
  unconfirmedCalls: number;
  inputTokens: number | null;
  outputTokens: number | null;
  durationMs: number | null;
  usageKnownCalls: number;
  usageMissingCalls: number;
  durationKnownCalls: number;
  networkAttempts: number;
  failedNetworkAttempts: number;
  unconfirmedNetworkAttempts: number;
  usageCoverage: "complete" | "partial" | "none";
};
export type ModelUsage = {
  runId: string | null;
  runStatus: string | null;
  date: string | null;
  timezone: string | null;
  windowStart: string | null;
  windowEnd: string | null;
  asOf: string;
  runDurationMs: number | null;
  summary: ModelUsageSummary;
  models: {
    resourceId: string;
    modelId: string;
    apiStyle: string;
    summary: ModelUsageSummary;
  }[];
};

File: backend/app/schemas/model_usage.py
"""Reported model usage with explicit coverage, never provider billing estimates."""

from datetime import date as CalendarDate
from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.common import CamelModel


class ModelUsageSummary(CamelModel):
    model_calls: int = 0
    confirmed_calls: int = 0
    failed_calls: int = 0
    unconfirmed_calls: int = 0
    input_tokens: int | None = None
    output_tokens: int | None = None
    duration_ms: int | None = None
    usage_known_calls: int = 0
    usage_missing_calls: int = 0
    duration_known_calls: int = 0
    network_attempts: int = 0
    failed_network_attempts: int = 0
    unconfirmed_network_attempts: int = 0
    usage_coverage: Literal["complete", "partial", "none"] = "none"


class ModelUsageBreakdown(CamelModel):
    resource_id: str
    model_id: str
    api_style: str
    summary: ModelUsageSummary


class ModelUsageRead(CamelModel):
    run_id: str | None = None
    run_status: str | None = None
    date: CalendarDate | None = None
    timezone: str | None = None
    window_start: datetime | None = None
    window_end: datetime | None = None
    as_of: datetime
    run_duration_ms: int | None = None
    summary: ModelUsageSummary
    models: list[ModelUsageBreakdown] = Field(default_factory=list)
